examples/advanced/llm_hf/README.md

```diff
@@ -0,0 +1,137 @@
+```
+
+The PEFT curves are shown below, blue for single call, black for iterative. We can see the "zig-zag" pattern in the iterative training loss curve.
+
+![peft](./figs/cen_peft.png)
+
+Similar patterns can be observed from the SFT curves
+
+![sft](./figs/cen_sft.png)
+
+### Adaptation Step 2: federated with NVFlare
+Once we have the iterative training script ready with "starting model" loading capability, it can be easily adapted to a NVFlare trainer by using [Client API](../../hello-world/ml-to-fl/pt/README.md).
+
+The major code modifications are for receiving and returning the global model (replacing the constant one used by iterative training), as shown below:
+
+![diff](./figs/diff_fl_1.png)
+![diff](./figs/diff_fl_2.png)
+
+## Job for NVFlare FL Training
+With the local training script ready, we can go ahead to generate the NVFlare job configs by reusing the job templates from [sag_pt](../../../job_templates/sag_pt/).
+
+Let's set the job template path with the following command.
+```bash
+nvflare config -jt ../../../job_templates/
+```
+Then we can check the available templates with the following command.
+```bash
+nvflare job list_templates
+```
+We can see the "sag_pt" template is available, with which we further generate job configs for SFT and PEFT as:
+```
+nvflare job create -force -j "./jobs/hf_peft" -w "sag_pt" -sd "code" \
+  -f meta.conf min_clients=1 \
+  -f config_fed_client.conf app_script="hf_sft_peft_fl.py" app_config="--model_path ${PWD}/model/Llama-2-7b-hf --data_path_train ${PWD}/dataset/dolly/training.jsonl --data_path_valid ${PWD}/dataset/dolly/validation.jsonl --output_path llama2-7b-dolly-peft --mode 1" \
+  -f config_fed_server.conf model_class_path="hf_peft_model.CausalLMPEFTModel" components[0].args.model.args.model_path="${PWD}/model/Llama-2-7b-hf" min_clients=1 num_rounds=3 key_metric="eval_loss" negate_key_metric=True
+```
+and
+```
+nvflare job create -force -j "./jobs/hf_sft" -w "sag_pt" -sd "code" \
+  -f meta.conf min_clients=1 \
+  -f config_fed_client.conf app_script="hf_sft_peft_fl.py" app_config="--model_path ${PWD}/model/Llama-2-7b-hf --data_path_train ${PWD}/dataset/dolly/training.jsonl --data_path_valid ${PWD}/dataset/dolly/validation.jsonl --output_path llama2-7b-dolly-sft --mode 0" \
+  -f config_fed_server.conf model_class_path="hf_sft_model.CausalLMModel" components[0].args.model.args.model_path="${PWD}/model/Llama-2-7b-hf" min_clients=1 num_rounds=3 key_metric="eval_loss" negate_key_metric=True  
+```
+
+For both client and server configs, we only set the necessary parameters for the SFT and PEFT tasks, and leave the rest to the default values.
+
+## Federated Training
+With the produced job, we run the federated training on a single client using NVFlare Simulator.
+```
+nvflare simulator -w ./workspace_fl/hf_peft -n 1 -t 1 ./jobs/hf_peft
+```
+and
+```
+nvflare simulator -w ./workspace_fl/hf_sft -n 1 -t 1 ./jobs/hf_sft
+``` 
+
+## Results
```

Comment:
How about FedAvg with more than one client?

Reply:
this example shows the "capability" of nvflare handling llm training,  
meta-parameter tuning will be needed to the training scheme to show better performance from say, a 3-client fedavg (like what we show for the megatro-1.3b), but current results do not have the evidence yet